packages/react/src/components/TableHandles/ExtendButton/ExtendButton.tsx

```diff
@@ -0,0 +1,181 @@
+    props.freezeHandles();
+    setEditingState({
+      startPos: props.orientation === "row" ? event.clientX : event.clientY,
+      numOriginalCells:
+        props.orientation === "row"
+          ? props.block.content.rows[0].cells.length
+          : props.block.content.rows.length,
+      clickOnly: true,
+    });
+  };
+
+  // Extends columns/rows on when moving the mouse.
+  useEffect(() => {
+    const callback = (event: MouseEvent) => {
+      if (editingState === null) {
+        return;
+      }
+
+      const diff =
+        (props.orientation === "row" ? event.clientX : event.clientY) -
+        editingState.startPos;
+
+      const numCells =
+        editingState.numOriginalCells +
+        Math.floor(diff / (props.orientation === "row" ? 100 : 31));
+      const block = props.editor.getBlock(props.block)!;
+      const numCurrentCells =
+        props.orientation === "row"
+          ? block.content.rows[0].cells.length
+          : block.content.rows.length;
+
+      if (
+        editingState.numOriginalCells <= numCells &&
+        numCells !== numCurrentCells
+      ) {
+        props.editor.updateBlock(props.block, {
+          type: "table",
+          content:
+            props.orientation === "row"
+              ? getContentWithAddedCols(
+                  props.block.content,
+                  numCells - editingState.numOriginalCells
+                )
+              : getContentWithAddedRows(
+                  props.block.content,
+                  numCells - editingState.numOriginalCells
+                ),
+        });
+        // Edge case for updating block content as `updateBlock` causes the
+        // selection to move into the next block, so we have to set it back.
+        if (block.content) {
+          props.editor.setTextCursorPosition(props.block);
+        }
+        setEditingState({ ...editingState, clickOnly: false });
+      }
+    };
+
+    document.body.addEventListener("mousemove", callback);
+
+    return () => {
+      document.body.removeEventListener("mousemove", callback);
+    };
+  }, [editingState, props.block, props.editor, props.orientation]);
+
+  // Stops mouse movements from extending columns/rows when the mouse is
+  // released. Also extends columns/rows by 1 if the mouse wasn't moved enough
+  // to add any, imitating a click.
+  useEffect(() => {
+    const callback = () => {
+      if (editingState?.clickOnly) {
+        props.editor.updateBlock(props.block, {
+          type: "table",
+          content:
+            props.orientation === "row"
+              ? getContentWithAddedCols(props.block.content)
+              : getContentWithAddedRows(props.block.content),
+        });
+      }
+
+      setEditingState(null);
+      props.unfreezeHandles();
+    };
+
+    document.body.addEventListener("mouseup", callback);
+
+    return () => {
+      document.body.removeEventListener("mouseup", callback);
+    };
+  }, [
+    editingState?.clickOnly,
```

Comment:
I don't think we have a choice here because we need the `mousemove` and `mouseup` handlers to be attached to the document body, because the mouse cursor is not always above the extend button while dragging. This is also why we can't use a click event, as those only fire when the same element is hovered on both the mouse down and mouse up.